iotileship/iotile/ship/resources/filesystem_manager.py

```diff
@@ -0,0 +1,120 @@
+''' A Resource for operating on a Filesystem
+
+    Can work on either a directory or a unmounted block device
+
+    Important: mounting and unmounting of a block device requires the udisksctl utility, which is included with
+    most distributions of Ubuntu
+
+    All Steps using this Resource will use self.root as the root directory to find other files
+
+    Root directory is a pathlib.Path
+
+'''
+
+import subprocess
+import pathlib
+import sys
+
+from iotile.core.utilities.schema_verify import DictionaryVerifier
+from iotile.core.utilities.schema_verify import StringVerifier
+from iotile.core.exceptions import ArgumentError
+from .shared_resource import SharedResource
+
+RESOURCE_ARG_SCHEMA = DictionaryVerifier(desc="filesystem_manager arguments")
+RESOURCE_ARG_SCHEMA.add_required("res_path", StringVerifier("path to resource to work with. "
+                                                            "Specify file:// for a directory or block:// for an unmounted disk partition"))
+
+
+class FilesystemManagerResource(SharedResource):
+    '''
+    Arguments:
+        path (str): Required. One of:
+            file://<path> : Path to a directory that will be worked on
+            block://<path> : Path to a block device to work on (typically /dev/sd...)
+    '''
+
+
+    ARG_SCHEMA = RESOURCE_ARG_SCHEMA
+    _allowed_types = ["file", "block"]
+
+    def __init__(self, args):
+        super(FilesystemManagerResource, self).__init__()
+
+        self._path = args.get('res_path')
+
+        if not "://" in self._path:
+            raise ArgumentError("Cannot detect resource type (looking for something like 'file://)'",
+                                parameter_name="res_path")
+
+        self._type = self._path.split('://')[0]
+        self._resource_loc = self._path.split('://')[1]
+
+        if self._type not in self._allowed_types:
+            raise ArgumentError("FilesystemManagerResource unknown filesystem type", parameter_name="res_path")
+
+        if self._type == "block" and sys.platform != "linux":
+            raise ArgumentError("block;// paths are only supported on Linux. \
```

Comment:
`\` line continuations are not needed in python when you are inside of a `( )` because that implicitly continues across lines.

Reply:
This is the current thing I'm fighting with, as the atomic moves are ruining the permissions. You've also just reminded me that I forgot to document the permission stuff in the iotile ship repo.

The solution as currently implemented in meta-arch-systems is to generate files with a group with a specific gid, assign that group to the iotile user, and require that the user running the personalize recipe also be a member of a group with the same gid. 

In more detail:
I've got the daemons running as the iotile user, with the iotile user assigned to group iotile (specifically assigned to gid 19057). On my local machine, I made a group named "yoc_iotile" with gid 19057 and added my local account to the group. This gives my machine access to rewrite the file.  There's also the passwd file holding the passwd hash for the iotile user, it should be owned by root, technically, but assigning it to a group that is not assigned to any user should accomplish the same protections. It's writeable by gid 15666 and on my local machine I added my user to a group called "yoc_noaccess" with gid 15666. 

Group ID's are typically between 0 and 1000, with user's primary groups being automatically numbered at 1000 and decreasing (user 2's group is 999). Gids up to 60,000 are commonly supported, so I picked two arbitrary gids in that range to use for the iotile group and the noaccess group.

Reply:
@timothyfroehlich If you just file.write() does it maintain the permissions even though the file contents are truncated and overwrritten?  Then I would say that we just enforce that all files are precreated and `iotile-ship` only ever changes their contents.

Reply:
Also, yes, files will be owned by whoever last modifies the file. Changing the group should handle permissions issues. Changing owners would require the iotileship script be run with sudo which I'm very hesitant to require.

Reply:
and then we don't have to worry about matching `gid` or user_id.  I would drop all of the atomic writes then per my other comments since the main benefit of atomic writes is when you are running code unattended in a situation where power could be lost at any time.  That's not the situation when you're running `iotile-ship` offline, on an SD card, during manufacturing time.

Reply:
It doesn't seem like the file ownership is modified by a write operation, unless `root` is special:

```shell
timburke@eng2:~$ touch temp
timburke@eng2:~$ sudo -s
[sudo] password for timburke:
root@eng2:~# ls -l
-rw-rw-r-- 1 timburke timburke    0 Apr 16 14:41 temp
root@eng2:~# echo "hello" > temp
root@eng2:~# ls -l
-rw-rw-r-- 1 timburke timburke    6 Apr 16 14:42 temp
root@eng2:~# cat temp
hello
```

I created a file as `timburke` and wrote to it as root without affecting the owner or group.

Reply:
Yeah I just verified with another user that wasn't root.  Just doing a write (even a complete truncation) to file as long as you don't destroy it doesn't seem to affect the ownership on linux.

Reply:
So I believe I've identified an issue in arch-node-tasks either way. While the files were created by yocto with 664 permissions, the atomic-rewrite code in mqtt_bridge was recreating the files with 644 permissions, removing the group's write permissions. This caused subsequent changes to fail. I'm not 100% sure because I saw it happen once, but I think I'm going to be a lot happier if I just get everything to use the same code for atomic moving, which needs to include preserving the gid and permissions on the file. The atomicwrite code does not do this.

So here's a new idea, I could make a new atomic re-write that starts by shutil.copy (or copy2?) the file, which would preserve all permissions including owner. If that works it'd probably be best.

Reply:
Scratch that...
On POSIX platforms, this means that file owner and group are lost as well as ACLs. On Mac OS, the resource fork and other metadata are not used. This means that resources will be lost and file type and creator codes will not be correct. On Windows, file owners, ACLs and alternate data streams are not copied.

https://docs.python.org/3/library/shutil.html#shutil.chown

Reply:
Could just read the owner / group of the original file and chown the new one.

Reply:
It looks like `bindfs` can be used to obviate the need to have the group ownership of a file be numerically equivalent to the group of a user running the personalization step.  It seems like you can tell bindfs to ignore the actual file ownership bits during the bind mount so that you are not prohibited from modifying the files (provided you don't create new files):

https://unix.stackexchange.com/a/161709

That should solve the need to have a shared group between the SD card in question and your computer running `iotile-ship`.

Reply:
Here's a more complete example showing how to even map uid/gid using bindfs so you can create files with the correct uid/gid from any linux computer:
https://unix.stackexchange.com/a/158695

Reply:
The mounting part would need to be run as root, but if I go back to handle any block device mounting in a shell script, then I should be able to just run the required commands and not the whole thing. Or I can make a sudoers file to install that gives permission to do bind.

Reply:
I'm dragging bad today and a second espresso hasn't helped. I'll let this settle in my head and work on it more tonight.

Reply:
Understood.

To summarize, I would say:

- it looks like having this all work magically without needing someone to do something as root at somepoint isn't feasible because we are working with an external filesystem whose numerical ids are not going to match in general with what is on the user's computer.
- this makes the `block://` option not super useful right now unless it is able to elevate and run as root so that the user can actually edit the files that they are supposed to regardless of what the uid/gid on the actual fs are.
- using `bindfs` we can fix all of the uid/gid issues by making the user mount the fs first using bindfs to make all fixes writable by the current user.  
- then we make sure we don't create any new files (inadvertently via atomicwrites, for example) during the `iotile-ship` execution.

This means the user will need to bind mount the directory they want to work on prior to running iotile-ship but that's easy to document in a README